### PR TITLE
initial commit of source code

### PR DESCRIPTION
**Pull request by nchancel**
This is the initial commit for the code, we will have to make another commit to put the arXiv number in the readme (and later for the full reference when published), but otherwise it should be good to go.

### CODE REVIEW

File: Zeno_constraint/zeno_constraint.py [hunk 1]
```diff
@@ -0,0 +1,478 @@
...
+    for i_add in range(n_var): # loop through first index
+        H=H+h[i_add]*kron_ith_position(M,i_add,n_var) # add linear term
+        for j_add in range(i_add,n_var): # loop through second index
+            if not J[i_add,j_add]==0: # don't perform tensor products for terms which don't contribute anyway
+                H=H+J[i_add,j_add]*kron_ith_position(M,i_add,n_var)@kron_ith_position(M,j_add,n_var) # dot prodoct of tensor products
+    return H
+    
+def kron_ith_position_vec(V,i_pos,n_var): # uses tensor products to create a matrix with an identity at all positions except for the ith assumes all variables have the same size which is the size of M
+    '''
+        M is the vector to be place at position i_pos
```
Comment (qci-github) [inline Zeno_constraint/zeno_constraint.py:39]: M is different from the arg V.

File: Zeno_constraint/zeno_constraint.py [hunk 2]
```diff
@@ -0,0 +1,478 @@
...
+    H_single=rotate_disp_ham(theta) # imaginary Hamiltonian for a single angle
+    H_tot=sparse.csr_array((3**n_qubit,3**n_qubit)) # empty array
+    for i_pos in range(n_qubit): # empty starting matrix
+        H_tot=H_tot+kron_ith_position(H_single,i_pos,n_qubit) # add new matrix
+    return H_tot
+    
+def multi_qubit_measure_project_xi(theta,n_qubit): # creates a projector which projects out the xi state at each qubit
+    '''
+        theta is an angle used to rotate between an undefined state and the |+> state, theta=0 defines full rotation into the undefined state, theta= pi/2 is fully in |+>
+        n is the number of qubit variables, n.b. because of the undefined state, the size of the total matrix is 3^n_qubit
```
Comment (qci-github) [inline Zeno_constraint/zeno_constraint.py:75]: n -> n_qubit

File: Zeno_constraint/zeno_constraint.py [hunk 3]
```diff
@@ -0,0 +1,478 @@
...
+            print('zeno_sweep_H_expm_mult received an unrecognised method '+method+' returning a NaN value this is likely to cause errors downstream')
+            return np.nan
+            
+        data_list[i_step]=measure_func(state_vec) # measure desired quantities
+    return data_list
+
+def qubit_variable_pm_tilde_transform(theta): # Gives a single variable transform to the plus tilde minus tilde basis
+    '''
+        theta is the rotation angle for the projection in the three-state space
+        P is a matrix where the first column corresponds to + tilde and the second to - tilde in the 0, 1 basis
```
Comment (qci-github) [inline Zeno_constraint/zeno_constraint.py:268]: P not included in args

File: Zeno_constraint/zeno_constraint.py [hunk 4]
```diff
@@ -0,0 +1,478 @@
...
+    '''
+    P=np.array([[np.cos(theta),1],[np.cos(theta),-1]])/np.sqrt(2)
+    return P
+
+def project_to_allowed_states(allowed_state): # allowed states is a boolean vector of states which are allowed
+    '''
+        allowed_states is vector with zeros corresponding to non-allowed states
+        P is a matrix which projects from a larger set of states to a smaller set of allowed states
```
Comment (qci-github) [inline Zeno_constraint/zeno_constraint.py:276]: P not included in args

File: Zeno_constraint/adiabt_run_script.py [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+        method='eigh'
+        data_list=zc.zeno_sweep_H_expm_mult(H_s,t_tot,start_state,measure_func,method=method,E_V_precomp_list=E_V_precomp_list)
+        final_success_probs_nI2[i_const_str,i_anneal_time]=data_list[-1][1]
+        print(i_anneal_time)
+        sat_prob_list=[[]+[data[1]] for data in data_list]
+        #plt.plot(sat_prob_list)
+        #plt.show()
+        print('total sweep calculation time= '+str(time.time()-t_start))
+    np.save('adiabat_2_sweep_constrain.npy',{'t_anneal_list':t_anneal_list,'final_success_probs': final_success_probs_nI2,'constraint_str_list':str_list})
+    #plt.semilogx(str_list,final_success_probs_nI2)
```
Comment (qci-github) [inline Zeno_constraint/adiabt_run_script.py:59]: Consider a `render_plot` argument. (defaults to false)

**Review by qci-github — COMMENTED** [review]
LGTM.

File: Zeno_constraint/zeno_constraint.py (hunk 1 above)
Comment (nchancel) [reply Zeno_constraint/zeno_constraint.py:39]: Yes, should read V, this is an error from copying and pasting a comment block from another function

File: Zeno_constraint/zeno_constraint.py (hunk 2 above)
Comment (nchancel) [reply Zeno_constraint/zeno_constraint.py:75]: I think this was from a variable name change which was not propagated to the comments, but should be corrected as you point out

File: Zeno_constraint/zeno_constraint.py (hunk 4 above)
Comment (nchancel) [reply Zeno_constraint/zeno_constraint.py:276]: P is what the function returns should be made more clear

File: Zeno_constraint/zeno_constraint.py (hunk 3 above)
Comment (nchancel) [reply Zeno_constraint/zeno_constraint.py:268]: P is what the function returns should be made more clear

File: Zeno_constraint/adiabt_run_script.py (hunk 5 above)
Comment (nchancel) [reply Zeno_constraint/adiabt_run_script.py:59]: Good idea

**Comment by nchancel** [comment]
I've made changes locally and will make a new commit and PR with the implemented changes (I don't know if there is a more graceful way to do this).

**Pull request closed without merging**